javascript/ql/src/semmle/javascript/frameworks/Markdown.qll

```diff
@@ -2,159 +2,188 @@
-      exists(UnifiedChain chain |
-        // sanitizer. Mostly looking for `rehype-sanitize`, but also other plugins with `sanitize` in their name.
-        not chain.getAUsedPlugin().getALocalSource() =
-          DataFlow::moduleImport(any(string s | s.matches("%sanitize%")))
-      |
-        pred = chain.getInput() and
-        succ = chain.getOutput()
+      exists(DataFlow::CallNode call | call = DataFlow::moduleImport("snarkdown").getACall() |
+        call = succ and
+        pred = call.getArgument(0)
       )
     }
   }
-}
-
-/**
- * A taint step for the `snarkdown` library.
- */
-private class SnarkdownStep extends TaintTracking::SharedTaintStep {
-  override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
-    exists(DataFlow::CallNode call | call = DataFlow::moduleImport("snarkdown").getACall() |
-      call = succ and
-      pred = call.getArgument(0)
-    )
-  }
-}
 
-/**
- * Classes and predicates for modelling taint steps the `markdown-it` library.
- */
-private module MarkdownIt {
   /**
-   * The creation of a parser from `markdown-it`.
+   * Classes and predicates for modelling taint steps the `markdown-it` library.
    */
-  private API::Node markdownIt() {
-    exists(API::InvokeNode call |
-      call = API::moduleImport("markdown-it").getAnInvocation()
+  private module MarkdownIt {
+    /**
+     * The creation of a parser from `markdown-it`.
+     */
+    private API::Node markdownIt() {
+      exists(API::InvokeNode call |
+        call = API::moduleImport("markdown-it").getAnInvocation()
+        or
+        call = API::moduleImport("markdown-it").getMember("Markdown").getAnInvocation()
+      |
+        call.getParameter(0).getMember("html").getARhs().mayHaveBooleanValue(true) and
+        result = call.getReturn()
+      )
       or
-      call = API::moduleImport("markdown-it").getMember("Markdown").getAnInvocation()
-    |
-      call.getParameter(0).getMember("html").getARhs().mayHaveBooleanValue(true) and
-      result = call.getReturn()
-    )
-    or
-    exists(API::CallNode call |
-      call = markdownIt().getMember(["use", "set", "configure", "enable", "disable"]).getACall() and
-      result = call.getReturn() and
-      not call.getParameter(0).getAValueReachingRhs() =
-        DataFlow::moduleImport("markdown-it-sanitizer")
-    )
-  }
-
-  /**
-   * A taint step for the `markdown-it` library.
-   */
-  private class MarkdownItStep extends TaintTracking::SharedTaintStep {
-    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
       exists(API::CallNode call |
-        call = markdownIt().getMember(["render", "renderInline"]).getACall()
-      |
-        succ = call and
-        pred = call.getArgument(0)
+        call = markdownIt().getMember(["use", "set", "configure", "enable", "disable"]).getACall() and
+        result = call.getReturn() and
+        not call.getParameter(0).getAValueReachingRhs() =
+          DataFlow::moduleImport("markdown-it-sanitizer")
```

Comment:
Should this negation be in the `preservesHtml` predicate?

Reply:
I think that would require adding a `boolean isSanitized` parameter to the `markdownIt()` predicate.   

The `MarkdownStep::step` predicate is documented as being a taint-step. 
And if `markdown-it-sanitizer` is used, then there is no taint-step.  
So I think it's fine.